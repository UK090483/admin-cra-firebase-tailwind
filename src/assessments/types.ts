import { Record, Identifier } from "../types";
export type AssessmentRating = 1 | 2 | 3 | 4;

export type AssessmentStatus = "created" | "hidden";

export interface IAssessmentPoints {
  isInnovative?: AssessmentRating;
  isClearUSP?: AssessmentRating;
  isTherePotential?: AssessmentRating;
  doTheyKnowCompetitors?: AssessmentRating;
  isBusinessModelSolid?: AssessmentRating;
  areNumbersPromising?: AssessmentRating;
  isFundingSolid?: AssessmentRating;
  isTeamExperienced?: AssessmentRating;
  willFutureCitiesBenefit?: AssessmentRating;
  isComplianceWithSDG?: AssessmentRating;
  isPresentationGood?: AssessmentRating;
}

export interface IAssessmentRecord extends Record, IAssessmentPoints {
  application_id: Identifier;
  application_name: string;
  judge_id: Identifier;
  judge_name: string;
  isInnovativeReason?: string;
  isClearUSPReason?: string;
  isTherePotentialReason?: string;
  doTheyKnowCompetitorsReason?: string;
  isBusinessModelSolidReason?: string;
  areNumbersPromisingReason?: string;
  isFundingSolidReason?: string;
  isTeamExperiencedReason?: string;
  willFutureCitiesBenefitReason?: string;
  isComplianceWithSDGReason?: string;
  isPresentationGoodReason?: string;
  status: AssessmentStatus;
  sum?: number;
}
