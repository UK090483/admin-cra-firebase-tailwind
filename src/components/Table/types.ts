import * as React from "react";

export interface IColumn {
  field: string;
  use: string;
  use_in_search?: boolean;
  use_in_display?: boolean;
  width?: string;
  render?: (row: IRow) => JSX.Element | string | number;
}
export interface IRowData {
  [name: string]: any;
}

export interface IBulkActionsProps {
  bulkItems: IBulkState;
  data: IRow[];
  setBulkItems: (bulkItems: IBulkState) => void;
}

export interface ITableOptions {
  bulkActions?: (props: IBulkActionsProps) => JSX.Element;
  showFilter?: boolean;
  fixedFilter?: (
    filter: ITableFilterState | null,
    setFilter: (filter: ITableFilterState | null) => void,
    columns: IColumn[],
    rows: IRow[]
  ) => JSX.Element;
}

export interface ITableProps {
  name?: string;
  columns: IColumn[];
  rows: IRow[];
  per_page?: number;
  searchFields?: string[];
  onRowClick?: IRowClickFunction;
  options?: ITableOptions;
  rowStyle?: (row: IRow) => string | void;
}
export interface ITableSortState {
  field: string;
  direction: "desc" | "asc";
}

export interface IBulkState {
  [id: string]: boolean;
}

export type ITableFilter = [
  "===" | "!==" | ">" | "<",
  string | boolean | undefined
];
export interface ITableFilterState {
  [id: string]: ITableFilter[];
}

export interface IColumn {
  field: string;
  use: string;
  use_in_search?: boolean;
  use_in_display?: boolean;
  render?: (row: IRow) => JSX.Element | string | number;
}

export interface ITableColumns extends Array<IColumn> {}

export interface IRow {
  [key: string]: any;
}

export type IRowClickFunction = (data: any) => void;
export declare type RowArray = IRow[];
export declare type renderFunction = (
  row: IRow,
  col: IColumn,
  display_value: string
) => JSX.Element;
export interface Iprop {
  rows: IRow[];
  columns: IColumn[];
  row_render?: renderFunction;
  per_page?: number;
  no_content_text?: string;
  debounce_search?: number;
  table_header?: string;
  show_search?: boolean;
  onRowClick?: IRowClickFunction;
}
/**
 * Interface describing data after the active page has been set
 * @props pagination: Has the full paginated data for the rows prop
 * @props back_button_clickable: indicates if back button is clickable
 * @props forward_button_clickable: Dictates if the forward button is clickable
 * @props current_rows: The rows to be shown in the table(The current page rows)
 * @props active_page_number: The currently selected page number
 * @param {string} search_string: The Text used in the search field
 *
 */
export interface ItableState {
  pagination: ItableLinks;
  back_button_clickable: boolean;
  forward_button_clickable: boolean;
  current_rows: IRow[];
  active_page_number: number;
  search_string: string;
}
export interface IpageNum {
  [key: number]: {
    page_row_array: IRow[];
    back_button_clickable: boolean;
    forward_button_clickable: boolean;
    is_active: boolean;
  };
}
export interface ItableLinks {
  page_number_store: number[];
  page_map: IpageNum;
  all_rows: IRow[];
}
export interface InavComponProp {
  back_button_clickable: boolean;
  forward_button_clickable: boolean;
}
export interface IpageActiveSet {
  table_links: ItableLinks;
  back_button_clickable: boolean;
  forward_button_clickable: boolean;
  current_rows: IRow[];
  active_page_number: number;
}
export interface IpaginateProps extends ItableState {
  backButtonOnclick: (event: React.MouseEvent) => void;
  forwardButtonOnclick: (event: React.MouseEvent) => void;
  pageNumberClick: (event: React.MouseEvent, page_number: number) => void;
}
export declare type stringObj = {
  [key: string]: any;
};
