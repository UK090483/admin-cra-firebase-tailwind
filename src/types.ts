import { Timestamp } from "@firebase/firestore-types";

export type Identifier = string;

export type FirebaseTimestamp = Timestamp;

export type RoutingParam = {
  id: Identifier;
};

export interface RecordBase {
  id: Identifier;
  created_at: Timestamp;
  created_by: Identifier;
  updated_at: Timestamp;
  updated_by: Identifier;
}

export type stage =
  | "first_filter"
  | "assign_prejudges"
  | "first_evaluation"
  | "final_evaluation";
