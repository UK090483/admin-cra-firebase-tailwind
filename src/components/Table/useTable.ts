import * as React from "react";
import tableFilter from "./helper/tableFilter";
import tableSort from "./helper/tableSort";
import tableSearch from "./helper/tableSearch";

import {
  ITableProps,
  ITableSortState,
  ITableFilterState,
  IBulkState,
} from "./types";

interface ITableState {
  page: number;
  filter: ITableFilterState | null;
  sort: ITableSortState | null;
  search: string | null;
  bulkItems: IBulkState;
  bulkMaster: boolean;
}

type Action =
  | { type: "setPage"; payload: number }
  | { type: "setFilter"; payload: ITableFilterState | null }
  | { type: "setSort"; payload: ITableSortState | null }
  | { type: "setState"; payload: ITableSortState }
  | { type: "setSearch"; payload: string }
  | { type: "setBulkItem"; payload: string }
  | { type: "setBulkItems"; payload: IBulkState };

type ITableReducer = (state: ITableState, action: Action) => ITableState;

const tableState: ITableState = {
  page: 0,
  filter: null,
  sort: null,
  search: null,
  bulkItems: {},
  bulkMaster: false,
};
const reducer: ITableReducer = (state, action) => {
  switch (action.type) {
    case "setPage":
      return {
        ...state,
        page: action.payload,
        bulkItems: {},
        bulkMaster: false,
      };
    case "setSort":
      return {
        ...state,
        sort: action.payload,
        bulkItems: {},
        bulkMaster: false,
      };
    case "setFilter":
      return {
        ...state,
        filter: action.payload,
        page: 0,
        bulkItems: {},
        bulkMaster: false,
      };

    case "setSearch":
      return {
        ...state,
        search: action.payload,
        page: 0,
        bulkItems: {},
        bulkMaster: false,
      };

    case "setState":
      return { ...state, ...action.payload };
    case "setBulkItem":
      return {
        ...state,
        bulkMaster: false,
        bulkItems: {
          ...state.bulkItems,
          [action.payload]: !!!state.bulkItems[action.payload],
        },
      };
    case "setBulkItems":
      return {
        ...state,
        bulkItems: action.payload,
        bulkMaster: !!Object.keys(action.payload).length,
      };
  }
};

const useTable = (props: ITableProps) => {
  const [state, dispatch] = React.useReducer(reducer, tableState);
  const { page, sort, filter, bulkItems, bulkMaster, search } = state;
  const {
    rows,
    columns,
    onRowClick,
    per_page = 10,
    options = {},
    searchFields = [],
  } = props;

  React.useEffect(() => {
    if (props.name) {
      const presistedState = window.localStorage.getItem(props.name);

      if (presistedState) {
        const ps = JSON.parse(presistedState);
        // console.log((ps.page = 0));
        // dispatch({ type: "setState", payload: ps });
      }
    }
  }, []);

  React.useEffect(() => {
    if (props.name) {
      window.localStorage.setItem(props.name, JSON.stringify(state));
    }
  }, [state]);

  const setPage = (page: number) => {
    dispatch({ type: "setPage", payload: page });
  };

  const setSort = (sort: ITableSortState | null) => {
    dispatch({ type: "setSort", payload: sort });
  };
  const setFilter = (filter: ITableFilterState | null) => {
    dispatch({ type: "setFilter", payload: filter });
  };
  const setBulkItem = (id: string) => {
    dispatch({ type: "setBulkItem", payload: id });
  };
  const setBulkItems = (bulkItems: IBulkState) => {
    dispatch({ type: "setBulkItems", payload: bulkItems });
  };

  const setSearch = (search: string) => {
    dispatch({ type: "setSearch", payload: search });
  };

  const data = React.useMemo(
    () =>
      tableSort(
        sort,
        tableFilter(filter, tableSearch(search, rows, searchFields))
      ),
    [rows, filter, sort, search]
  );

  const paginated = React.useMemo(
    () => data.slice(per_page * state.page, per_page * (state.page + 1)),
    [state.page, data]
  );

  const handleBulkMasterClick = (s: boolean) => {
    dispatch({
      type: "setBulkItems",
      payload: s
        ? paginated.reduce((acc, item) => ({ ...acc, [item.id]: true }), {})
        : {},
    });
  };

  const maxPage = Math.ceil(data.length / per_page);

  return {
    rows,
    data: paginated,
    page,
    setPage,
    maxPage,
    filter,
    setFilter,
    setSearch,
    search,
    columns,
    options,
    setSort,
    sort,
    hasBulk: !!options.bulkActions,
    bulkItems,
    onRowClick,
    setBulkItem,
    bulkMaster,
    handleBulkMasterClick,
    setBulkItems,
  };
};

export default useTable;
