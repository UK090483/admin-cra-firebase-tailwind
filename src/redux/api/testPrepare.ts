//  @jest-environment node
import { db, firebase } from "misc/firebase";

const buildTestCollection = async () => {
  const testArray = new Array(10);
  testArray.fill(1);
  return Promise.all(
    testArray.map((item, index) =>
      db
        .collection("testCollection")
        .doc("item" + index)
        .set({
          name: index,
          updated_at: firebase.firestore.Timestamp.now().toDate(),
        })
    )
  );
};

interface item {
  id: string;
}

const deleteTestCollection = async () => {
  const erasers: any[] = [];
  try {
    const collection = await db.collection("testCollection").get();
    collection.forEach((item: item) => {
      erasers.push(db.collection("testCollection").doc(item.id).delete());
    });
  } catch (error) {
    console.log(error);
  }

  return true;
  return Promise.all(erasers);
};

export { buildTestCollection, deleteTestCollection, db };
