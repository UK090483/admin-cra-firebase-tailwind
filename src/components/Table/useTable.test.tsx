/**
 * @jest-environment jsdom
 */

import useTable from "./useTable";
import { testColumns, testRows, testRows88 } from "./tests/testPrepare";
import { renderHook, act } from "@testing-library/react-hooks";
import { ITableProps } from "./types";

const CustomRender = (props: ITableProps) => {
  return renderHook(() => useTable({ ...props }));
};

describe("useTable", () => {
  test("should have correct initial state ", () => {
    const { result } = CustomRender({ columns: testColumns, rows: testRows });

    expect(result.current.page).toBe(0);
    expect(result.current.setPage).toBeTruthy();
    expect(result.current.maxPage).toBe(1);
    expect(result.current.data).toMatchObject(testRows);
    expect(result.current.filter).toBeNull;
    expect(result.current.setFilter).toBeTruthy();
    expect(result.current.columns).toMatchObject(testColumns);
    expect(result.current.options).toMatchObject({});
    expect(result.current.sort).toBeNull();
    expect(result.current.setSort).toBeTruthy();
    expect(result.current.onRowClick).toBeUndefined();
    expect(result.current.setBulkItem).toBeTruthy();
  });

  test("should have correct initial state  long List ", () => {
    const { result } = CustomRender({ columns: testColumns, rows: testRows88 });

    expect(result.current.page).toBe(0);
    expect(result.current.setPage).toBeTruthy();
    expect(result.current.maxPage).toBe(9);
    expect(result.current.data).toMatchObject(testRows88.slice(0, 10));
    expect(result.current.filter).toBeNull;
    expect(result.current.setFilter).toBeTruthy();
    expect(result.current.columns).toMatchObject(testColumns);
    expect(result.current.options).toMatchObject({});
    expect(result.current.sort).toBeNull();
    expect(result.current.setSort).toBeTruthy();
    expect(result.current.onRowClick).toBeUndefined();
    expect(result.current.setBulkItem).toBeTruthy();
  });

  test("should have working Pagination", () => {
    const { result } = CustomRender({ columns: testColumns, rows: testRows88 });

    expect(result.current.page).toBe(0);
    expect(result.current.data).toMatchObject(testRows88.slice(0, 10));
    act(() => {
      result.current.setPage(2);
    });
    expect(result.current.data).toMatchObject(testRows88.slice(20, 30));
    act(() => {
      result.current.setPage(4);
    });
    expect(result.current.data).toMatchObject(testRows88.slice(40, 50));
  });

  test("should handle BulkActions", () => {
    const { result } = CustomRender({ columns: testColumns, rows: testRows88 });
    act(() => {
      result.current.setBulkItem("2");
    });
    expect(result.current.bulkItems).toStrictEqual({ "2": true });
    act(() => {
      result.current.setBulkItem("2");
    });
    expect(result.current.bulkItems).toStrictEqual({ "2": false });
  });

  test("should handle Sort", () => {
    const { result } = CustomRender({ columns: testColumns, rows: testRows88 });

    act(() => {
      result.current.setSort({ field: "name", direction: "asc" });
    });

    expect(result.current.sort).toStrictEqual({
      field: "name",
      direction: "asc",
    });
  });

  test("should handle Filter", () => {
    const { result } = CustomRender({ columns: testColumns, rows: testRows88 });
    act(() => {
      result.current.setFilter({ name: [["===", "a"]] });
    });
    expect(result.current.filter).toStrictEqual({ name: [["===", "a"]] });
  });

  test("should handle Filter", () => {
    const { result } = CustomRender({ columns: testColumns, rows: testRows88 });
    act(() => {
      result.current.setFilter({ name: [["===", "a"]] });
    });
    expect(result.current.filter).toStrictEqual({ name: [["===", "a"]] });
  });

  test("should handle BulkMasterClick", () => {
    const { result } = CustomRender({ columns: testColumns, rows: testRows });
    act(() => {
      result.current.handleBulkMasterClick(true);
    });
    expect(result.current.bulkItems).toStrictEqual({
      1: true,
      2: true,
      3: true,
      4: true,
    });

    act(() => {
      result.current.handleBulkMasterClick(false);
    });

    expect(result.current.bulkItems).toStrictEqual({});
  });
});
