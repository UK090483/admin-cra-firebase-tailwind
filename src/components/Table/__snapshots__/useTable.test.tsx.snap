// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useTable should have correct initial state  1`] = `
Array [
  Object {
    "age": 40,
    "id": 4,
    "name": "d",
  },
  Object {
    "age": 30,
    "id": 3,
    "name": "c",
  },
  Object {
    "age": 20,
    "id": 2,
    "name": "b",
  },
  Object {
    "age": 10,
    "id": 1,
    "name": "a",
  },
]
`;
